llvm: make targets a multivalued variant

haampie opened the pull request:
This allow one to have two packages that depend on say `llvm targets=NVPTX` and
`llvm targets=AMDGPU` which Spack would merge into `targets=NVPTX,AMDGPU`.

The default is `targets=auto`, which corresponds with
`LLVM_TARGETS_TO_BUILD=all`, and gives a reasonable user experience when just
running `spack install llvm+clang`.

With this change, Spack will always add LLVM's host arch as an enabled target,
so even if you don't specify `targets=X86`, if your Spack target arch is say
`znver2`, it will add `X86` to `LLVM_TARGETS_TO_BUILD`. This is mostly so that
the user doesn't have to do the mapping from spack -> LLVM target, and can just
depend on the "special" targets.

haampie committed:
llvm: make targets a multivalued variant

haampie commented:
What do we think about composability issues? e.g. `targets=auto` does not compose with `targets=X86,WebAssembly`

sethrj commented on var/spack/repos/builtin/packages/iwyu/package.py:
```diff
@@ -33,14 +33,14 @@ class Iwyu(CMakePackage):
-    depends_on('llvm+all_targets', when='target=ppc64le:')
-    depends_on('llvm+all_targets', when='target=sparc:')
-    depends_on('llvm+all_targets', when='target=sparc64:')
+    depends_on('llvm', when='target=aarch64:')
+    depends_on('llvm', when='target=arm:')
+    depends_on('llvm', when='target=ppc:')
+    depends_on('llvm', when='target=ppcle:')
+    depends_on('llvm', when='target=ppc64:')
+    depends_on('llvm', when='target=ppc64le:')
+    depends_on('llvm', when='target=sparc:')
+    depends_on('llvm', when='target=sparc64:')
```
I think these all need to be `targets=x86` since IWYU needs to build x86-related code on non-x86 platforms.

sethrj replied:
@haampie This assumes the llvm dependency is `all` or `x86`, but for correctness shouldn't it include that explicitly?

sethrj commented on var/spack/repos/builtin/packages/llvm/package.py:
```diff
@@ -138,10 +138,15 @@ class Llvm(CMakePackage, CudaPackage):
+        values=auto_or_any_combination_of(
+            "AArch64", "AMDGPU", "ARM", "AVR", "BPF", "CppBackend", "Hexagon", "Lanai",
+            "Mips", "MSP430", "NVPTX", "PowerPC", "RISCV", "Sparc", "SystemZ",
+            "WebAssembly", "X86", "XCore"
```
Aren't variant values usually lowercase with underscores? It could be confusing when we have a spack arch `aarch64` but have to remember that llvm's usage is `AArch64`. Maybe we need to add standardized variant *values* to https://github.com/spack/spack/pull/25841 as well. But for this particular case, I would try to use Spack arch names  (e.g. `ppc`) and map those to the LLVM naming (`PowerPC`).

haampie replied:
There's no 1-to-1 mapping between spack targets and llvm targets though, so I'm not sure if we should spack'ify llvm arguments.

trws replied:
As a compromise, perhaps we can make it case insensitive for known targets?  I would always default to writing all upper or all lower, have tripped over that on some of the targets more than once.

haampie replied:
It's lowercase now

trws commented:
Why the `auto` option as a synonym for `all`?  If we're going to have both I would have expected it to work like the old default since `all` can be used for the new behavior.

trws commented:
> What do we think about composability issues? e.g. `targets=auto` does not compose with `targets=X86,WebAssembly`

It's unfortunate, but I don't know a way to fix it.  Unless there's a way to define a function that gets used to normalize multi-valued variants before hashing, not sure there's anything we can do other than always requiring users to list everything, which is not necessarily better.  @sethrj, @becker33 I'm not sure I'm up to date on interfaces, is there anything we can do about this at the package level short of making everyone list everything out every time?

sethrj commented:
ping @haampie  -- looks like this has gone a bit stale.

haampie committed:
Fix the targets variant values

1. Make them lowercase and add a mapping to cmake equivalent
2. auto -> all
2. Restore composability by using a multivalued variant, so that
   `targets=all` and `targets=x86` is combined to `targets=all,x86`
   which is then transformed into LLVM_TARGETS_TO_BUILD=all.

haampie commented:
I've pushed a new commit, hoping this makes everyone happy:

1. Make targets variant values lowercase and add a mapping to
    their cmake equivalent
2. `auto` is renamed to `all`
3. Restore composability by using a multivalued variant (not any combination of auto/[values...]), so that
    constraints `targets=all` and `targets=x86` are combined to
   `targets=all,x86` which is then transformed into
    `LLVM_TARGETS_TO_BUILD=all`

Regarding 3 there is an issue that `targets=all,x86` and `targets=all` are equivalent, but have a different hash, but well... I can live with that.

sethrj commented on var/spack/repos/builtin/packages/llvm/package.py:
```diff
@@ -138,10 +138,14 @@ class Llvm(CMakePackage, CudaPackage):
-        description="Build all supported targets, default targets "
-        "<current arch>,NVPTX,AMDGPU,CppBackend",
+        "targets",
+        default="all",
```
Do you have any idea of how changing from native-only to all-targets affects build time/exe size?

haampie replied:
I just checked that targets=all is 7000 tasks and targets=x86 is 6000 tasks according to ninja when leaving everything else as default. Not sure what happens if also clang and other variants are turned off

trws replied:
Why changing the default to 'all' here? It's certainly possible, but for some of the non-clang tools can have requirements on the system.  Recently had an error caused by a missing arm header building lldb for example, and it increases compile time non-trivially as well.  Not saying "no" mind you, but I'd like to know the reasoning.

haampie replied:
Sorrry, that was not intentional. Now the default is `nvptx,amdgpu` like it was.

haampie replied:
and by seth's comment: the default is now `none`.

sethrj commented on var/spack/repos/builtin/packages/zig/package.py:
```diff
@@ -17,6 +17,6 @@ class Zig(CMakePackage):
     )
 
-    depends_on('llvm@11.0.0: +all_targets')
+    depends_on('llvm@11.0.0:')
```
I would guess this also needs to be `targets=all`?

haampie replied:
Hm, not a Zig user, but I would assume most people would just want to build for native only, otherwise they can always to zig ^llvm targets=all?

sethrj replied:
I'm not one either, but potentially there's code in there that assumes the upstream LLVM has been build with all targets? Pinging maintainer @alalazo but the easy solution would just to be to add `targets=all` : it has the same meaning as the previous requirement and since the new default is `targets=all` it won't change the llvm concretization.

trws replied:
Zig is a bit of a special case in this respect, one of its core functions is being able to target any platform from any platform, [news piece here with some info](https://zig.news/kristoff/cross-compile-a-c-c-project-with-zig-3599).  I'd guess it may actually cause runtime failures, or at least a jarring experience, if that went away.

```suggestion
    depends_on('llvm@11.0.0: targets=all')
```

haampie committed:
use targets=x86 in iwyu

haampie committed:
Default to nvptx/amdgpu/host arch targets

haampie commented:
@sethrj / @trws should be ready, again

sethrj reviewed (dismissed):
A little weird that the gpu targets are on by default but you're right that it's the current behavior.

haampie commented:
No clue what other value to provide, since empty array is not allowed last time i checked.

Oh wait, maybe `none`?

haampie committed:
default to none

haampie commented:
Default is now `targets=none`.

I also think we could add a special type of variant that expands `all` to the values `amdgpu,...`, but that's for another pr.

trws commented:
@haampie, I'm happy with this except the zig thing.  Wanted to double-check my memory before saying anything about it but that community almost certainly relies on having all architectures in there.

haampie committed:
Update var/spack/repos/builtin/packages/zig/package.py

Co-authored-by: Tom Scogland <tom.scogland@gmail.com>

haampie commented:
Thanks for the reviews!

ye-luo commented:
> A little weird that the gpu targets are on by default but you're right that it's the current behavior.

It is not weird due to the fact that building and running are two separate things. I build executables with OpenMP offload to GPUs (using GPU backends) on a high core count CPU-only node and run the app on a GPU compute node.

Proposed the default to `all` change in https://github.com/spack/spack/pull/36675
